File: CS2500/Project 2/python/Sort.py
'''
# @file   Sort.py
# @author Evan Wilcox, CS2500, Section 1A
# @brief  A wrapper class for Merge Sort, Heap Sort and Quick Sort.
''' 

'''
# @class Sort
# @brief A class wrapper for Merge Sort and Insertion Sort.
''' 
class Sort:
    '''
    # @fn    __init__
    # @brief Class initializer.
    # @pre   No pre-condition.
    # @post  A Sort object is created.
    ''' 
    def __init__(self):
        self.debug = False      # indicates whether debug mode is enabled.
    

    '''
    # @fn    setDebug
    # @brief Sets the debug mode on or off.
    # @pre   val must be a boolean
    # @post  The new debug mode is set.
    # @param val new intended debug mode
    ''' 
    def setDebug(self, val):
        # Precondition
        if self.debug:
            assert type(val) == bool
        
        self.debug = val
    
    '''
    # @fn     MergeSort
    # @brief  Sorts the given array using Merge Sort.
    # @pre    A must be an array of integers.
    # @post   A is sorted.
    # @param  A an array containing integers.
    # @return A sorted array of integers.
    '''
    def MergeSort(self, A):
        # Precondition
        if self.debug:
            assert type(A) == list
            for i in range(len(A)):
                assert type(A[i]) == int
        
        if len(A) <= 1:
            return A
        else:
            m = len(A)//2
            L = self.MergeSort(A[:m])
            R = self.MergeSort(A[m:])
            
            # Postcondition
            if self.debug:
                assert self.checkSort(self.merge(L, R))
            
            return self.merge(L, R)

    '''
    # @fn     merge
    # @brief  Merges two sorted arrays in to one sorted array.
    # @pre    Both L and R must be arrays of integers in sorted order.
    # @post   L and R are merged in to one sorted array.
    # @param  L An array of sorted integers.
    # @param  R An array of sorted integers.
    # @return An array of sorted integers.
    ''' 
    def merge(self, L, R):
        # Preconditions
        if self.debug:
            assert type(L) == list
            assert type(R) == list
            
            for i in range(len(L)):
                assert type(L[i]) == int
                
            for i in range(len(R)):
                assert type(R[i]) == int
        
            assert self.checkSort(L)
            assert self.checkSort(R)

        S = []
        '''
        # loop description while loop that merges two arrays 
        # invariant:       S consists of elements from L and R in sorted order
        # proof:           
        # initialization:
        #    S is initialized to [] so it is sorted.
        #
        # maintenance: 
        #   The first element is removed from L or R and added to
        #   S. Because L and R are sorted S stays sorted.
        #
        # termination:
        #   Because an element is always removed from L or R either
        #   L or R size will equal 0 so the while loop will terminate.
        ''' 
        while len(L) > 0 and len(R) > 0:
            # Invariant
            if self.debug:
                assert self.checkSort(S)
            
            if L[0] < R[0]:
                S.append(L[0])
                L = L[1:]
            else:
                S.append(R[0])
                R = R[1:]
        
        if len(L) == 0:
            S = S + R
        elif len(R) == 0:
            S = S + L
        
        # Postcondition
        if self.debug:
            assert self.checkSort(S)
        
        return S


    def HeapSort(self, A):
        # Preconditions
        if self.debug:
            assert type(A) == list
            for i in range(len(A)):
                assert type(A[i]) == int
        
        heapSize = len(A)

        for i in range(heapSize//2, -1, -1):
            self.heapify(A, heapSize, i)
        
        for i in range(heapSize-1, 0, -1):
            A[i], A[0] = A[0], A[i]
            self.heapify(A, i, 0)
        
        # Postcondition
        if self.debug:
            assert self.checkSort(A)
        
        return A


    def heapify(self, A, heapSize, i):
        # Preconditions
        if self.debug:
            assert type(A) == list
            for j in range(len(A)):
                assert type(A[j]) == int

            assert type(heapSize) == int
            assert type(i) == int
        
        
        largest = i
        l = (2 * i) + 1    
        r = (2 * i) + 2

        if l < heapSize and A[l] > A[i]:
            largest = l
        if r < heapSize and A[r] > A[largest]:
            largest = r
        if largest != i:
            A[i], A[largest] = A[largest], A[i]
            self.heapify(A, heapSize, largest)

            # Postcondition
            if self.debug:
                assert self.checkHeap(A, heapSize, largest)


    def QuickSort(self, A):

        if len(A) == 0:
            return []
        else:
            p = self.partition(A)
            return self.QuickSort(A[:p]) + [A[p]] + self.QuickSort(A[p+1:])

    
    def partition(self, A): 
        low = 0
        high = len(A) - 1
        
        x = A[high]
        i = low-1
    
        for j in range(low, high): 
            if A[j] <= x: 
                i = i+1 
                A[i],A[j] = A[j],A[i] 
    
        A[i+1],A[high] = A[high],A[i+1] 
        return i+1
    

    '''
    # @fn     checkSort
    # @brief  Checks if the given array A is in sorted order.
    # @pre    A must be an array on integers.
    # @post   A boolean is returned.
    # @param  A an array of integers.
    # @return Boolean indicating if the passed array is sorted in 
              increasing order.
    ''' 
    def checkSort(self, A):
        # Preconditions
        if self.debug:
            assert type(A) == list
            for i in range(len(A)):
                assert type(A[i]) == int
        
        for i in range(len(A)-1):
            if A[i] > A[i+1]:
                return False

        return True


    def checkHeap(self, A, heapSize, i):
        # Preconditions
        if self.debug:
            assert type(A) == list
            assert type(heapSize) == int
            for j in range(len(A)):
                assert type(A[j]) == int

        
        if (i * 2 + 1) < heapSize:
            if A[i * 2 + 1] > A[i]:
                return False
        
        if (i * 2 + 2) < heapSize:
            if A[i * 2 + 2] > A[i]:
                return False
        
        return True

File: CS2500/Project 2/c++/Sort.cpp
////////////////////////////////////////////////////
// @file   Sort.cpp
// @author Evan Wilcox
// @brief  Function definitions for the Sort class
////////////////////////////////////////////////////

#include "Sort.h"

void Sort::swap(int *a, int *b)
{
  int temp = *a;
  *a = *b;
  *b = temp;
}


bool Sort::checkSort(int arr[], int l, int r)
{
  // Precondition
  if(debug)
  {
    assert(0 <= l);
    assert(l <= r);
  }
  
  for(int i = l; i < r-1; i++)
  {
    if(arr[i] > arr[i+1])
    {
      return false;
    }
  }

  return true;
}


bool Sort::checkHeap(int arr[], int heapSize, int i)
{
  // Precondition
  if(debug)
  {
    assert(0 <= heapSize);
    assert(0 <= i);
  }
  
  if((i * 2 + 1) < heapSize)
  {
    if(arr[i*2+1] > arr[i])
    {
      return false;
    }
  }

  if((i * 2 + 2) < heapSize)
  {
    if(arr[i*2+2] > arr[i])
    {
      return false;
    }
  }

  return true;
}


void Sort::merge(int arr[], int l, int m, int r)
{
  // Preconditions
  if(debug)
  {
    assert(0 <= l);
    assert(l <= m);
    assert(m <= r);
    assert(checkSort(arr, l, m));
    assert(checkSort(arr, m+1, r));
  }


  int i, j, k;
  int n1 = m - l + 1;
  int n2 = r - m;

  int *L = new int[n1];
  int *R = new int[n2];

  for(i = 0; i < n1; i++)
  {
    L[i] = arr[l+i];
  }
  for(j = 0; j < n2; j++)
  {
    R[j] = arr[m+1+j];
  }

  i = 0;
  j = 0;
  k = l;
  
  ////////////////////////////////////////////
  // merges the arrays L and R in to arr 
  // loop precondition:  arrays L and R are in sorted increasing order
  // loop postconsition: arr[l..r] is in increasing order
  // ivariant: arr[l..k] contains the smalest elements from 
  //           L and R in increasing order
  // proof:
  //    initialization: before the loop k=l so the subarry arr[l..k] 
  //                    has no elements so it is sorted
  //
  //    maintenance: either an element from L or R gets moved to arr
  //                 and k increases each iteration so arr[l..k] still 
  //                 contains elements from L and R
  //
  //    termination: at termination k = r so the subarry arr[l..r] contains 
  //                 the smallest elements from L and R at termination
  //
  ///////////////////////////////////////////
  while(i < n1 && j < n2)
  {
    // Invariant
    if(debug)
    {
      assert(checkSort(arr, l, k));
    }
    
    if(L[i] <= R[j])
    {
      arr[k] = L[i];
      i++;
    }
    else
    {
      arr[k] = R[j];
      j++;
    }
    k++;
  }

  while(i < n1)
  {
    arr[k] = L[i];
    i++;
    k++;
  }

  while(j < n2)
  {
    arr[k] = R[j];
    j++;
    k++;
  }

  delete L;
  delete R;

  // Postcondiion
  if(debug)
  {
    assert(checkSort(arr, l, r));
  }
}


void Sort::heapify(int arr[], int n, int i)
{
  // Preconditions
  if(debug)
  {
    assert(n >= 0);
    assert(i >= 0);
  }
  
  int largest = i;
  int l = 2*i + 1;
  int r = 2*i + 2;

  if(l < n && arr[l] > arr[largest])
  {
    largest = l;
  }

  if(r < n && arr[r] > arr[largest])
  {
    largest = r;
  }

  if(largest != i)
  {
    swap(&arr[i], &arr[largest]);
    heapify(arr, n, largest);
  }

  // Postcondition
  if(debug)
  {
    assert(checkHeap(arr, n, largest));
  }
}


int Sort::partition(int arr[], int l, int h)
{
  // Precondition
  if(debug)
  {
    assert(0 <= l);
    assert(l <= h);
  }
  
  int x = arr[h];
  int i = (l - 1);

  for(int j = l; j <= h-1; j++)
  {
    if(arr[j] <= x)
    {
      i++;
      swap(&arr[i], &arr[j]);
    }
  }

  swap(&arr[i+1], &arr[h]);
  return (i+1);
}


Sort::Sort()
{
  debug = false;
}


Sort::Sort(bool b)
{
  debug = b;
}


void Sort::setDebug(bool b)
{
  debug = b;
}


void Sort::print(int arr[], int n)
{
  // Precondition
  if(debug)
  {
    assert(0 <= n);
  }
  
  for(int i = 0; i < n-1; i++)
  {
    cout << arr[i] << ", ";
    if((i+1) % 10 == 0)
    {
      cout << endl;
    }
  }

  cout << arr[n-1] << endl << endl;
}


void Sort::mergeSort(int arr[], int l, int r)
{
  // Preconditions
  if(debug)
  {
    assert(0 <= l);
    assert(l <= r);
  }
  
  if(l < r)
  {
    int m = l+(r-l)/2;

    mergeSort(arr, l, m);
    mergeSort(arr, m+1, r);
    merge(arr, l, m, r);
  }

  // Postcondition
  if(debug)
  {
    assert(checkSort(arr, l, r));
  }
}


void Sort::heapSort(int arr[], int n)
{
  // Preconditions
  if(debug)
  {
    assert(0 <= n);
  }
  
  ////////////////////////////////////////////
  // this for loop builds a max heap
  // loop precondition:  n is the number of elements in arr
  // loop postconsition: arr is not a max heap
  // ivariant: the nodes i+1, i+2, .., n are roots of 
  //           a max heap
  // proof:
  //    initialization: i = n/2 so the nodes n/2 +1, n/2 +2,
  //                    ..., n are leafs so are trivial roots of
  //                    max heaps
  //
  //    maintenance: the children of i are numbered higher than i,
  //                 by the loop invariant they are both roots of max heaps
  //                 decrementing i restablishes the invariant
  //
  //    termination: at termination i = -1, each node 1, 2, 3 are 
  //                 roots of max heaps
  //
  ///////////////////////////////////////////
  for(int i = n/2 - 1;i >= 0; i--)
  {
    heapify(arr, n, i);

    // Invariant
    if(debug)
    {
      for(int j = n/2; j < n; j++)
      {
        assert(checkHeap(arr, n, j));
      }
    }
  }
  
  for(int i = n-1; i >= 0; i--)
  {
    swap(&arr[0], &arr[i]);
    heapify(arr, i, 0);
  }

  // Postcondition
  if(debug)
  {
    assert(checkSort(arr, 0, n));
  }
}


void Sort::quickSort(int arr[], int l, int h)
{
  // Preconditions
  if(debug)
  {
    assert(0 <= l);
    assert(l <= h);
  }
  
  if(l < h)
  {
    int p = partition(arr, l, h);
    quickSort(arr, l, p-1);
    quickSort(arr, p+1, h);
  }

  // Postcondition
  if(debug)
  {
    assert(checkSort(arr, l, h));
  }
}
